command-line-interface.md

```diff
@@ -0,0 +1,188 @@
+```sh
+$ funC [global-options] <COMMAND> [command-specific-options] <command-specific-arguments>
+```
+
+## Global options
+
+None are required, but the runtime MAY support options that start with at least one hyphen.
+Global options MAY take positional arguments (e.g. `--log-level debug`).
+Command names MUST NOT start with hyphens.
+The option parsing MUST be such that `funC <COMMAND>` is unambiguously an invocation of `<COMMAND>` (even for commands not specified in this document).
+If the runtime is invoked with an unrecognized command, it MUST exit with a nonzero exit code and MAY log a warning to stderr.
+
+## Character encodings
+
+This API specification does not cover character encodings, but runtimes SHOULD conform to their native operating system.
+For example, POSIX systems define [`LANG` and related environment variables][posix-lang] for [declaring][posix-locale-encoding] [locale-specific character encodings][posix-encoding], so a runtime in an `en_US.UTF-8` locale SHOULD write its [state](#state) to stdout in [UTF-8][].
+
+## Commands
+
+### create
+
+[Create][create] a container from a [bundle directory][bundle].
+
+* *Arguments*
+    * *`<ID>`* Set the container ID to create.
+* *Options*
+    * *`--bundle <PATH>`* Override the path to the [bundle directory][bundle] (defaults to the current working directory).
+    * *`--pid-file <PATH>`* The runtime MUST write the container PID to this path.
+* *Standard streams:*
+    * *stdin:* The runtime MUST attach its stdin directly to the container process without reading from it.
+    * *stdout:* The runtime MUST attach its stdout directly to the container process without writing to it.
+    * *stderr:* The runtime MUST attach its stderr to the container process, and MUST not write to it unless it exits with a non-zero code.
+* *Environment variables*
+    * *`LISTEN_FDS`:* The number of file descriptors passed.
+      For example, `LISTEN_FDS=2` would mean that the runtime MUST pass file descriptors 3 and 4 to the container process (in addition to the standard streams) to support [socket activation][systemd-listen-fds].
+* *Exit code:* Zero if the container was successfully created and non-zero on errors.
+
+Callers MAY block on this command's successful exit to trigger post-create activity.
+
+#### Example
+
+```
+# in a bundle directory with a process that echos "hello" and exits 42
+$ test -t 1 && echo 'stdout is a terminal'
+stdout is a terminal
+$ funC create hello-1
+$ echo $?
+0
+$ funC start hello-1
+hello
+$ echo $?
+0
+$ block-on-exit-and-collect-exit-code hello-1
+$ echo $?
+42
+$ funC delete hello-1
+$ echo $?
+0
+```
+
+`hello` shows up on in the terminal after `start`, because `start` happens to be called from the same shell session (using the same terminal) as `create`, and the container process inherited its standard streams from `create`.
+If you call `start` from a shell using a separate terminal, the container output would still have appeared in the `create` terminal.
+
```

Comment:
On Wed, Jul 27, 2016 at 07:37:38AM -0700, Mike Brown wrote:
> Runc: `create [--console value] <cid>` The --console option
> specifies the pty slave path for use with the created container.

My understanding after last week's meeting was that we were punting on
--console until @cyphar had time to do some research [1].  The current
description is just because the container process inherits the
‘create’ process's standard streams.  That is going to limit you to
one container per terminal unless you redirect the standard streams or
have containers prepared to handle SIGTTIN / SIGTTOU.

[1]: http://ircbot.wl.linuxfoundation.org/meetings/opencontainers/2016/opencontainers.2016-07-20-21.03.log.html#l-73